File: backends/php-laravel/app/Support/Exceptions/AppException.php
<?php

/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */

declare(strict_types=1);

namespace App\Support\Exceptions;

use RuntimeException;
use Symfony\Component\HttpFoundation\Response;

/**
 * アプリケーションExceptionクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 * @package App\Support\Exceptions
 */
class AppException extends RuntimeException
{
    /**
     * コンストラクタ。
     *
     * @param int $status HTTPステータスコード
     * @param string $key メッセージキー
     * @param array $replace メッセージ置換文字列
     */
    public function __construct(
        int $status,
        string $key,
        array $replace = [],
    ) {
        $message = __('validation.custom.' . $key, $replace);
        parent::__construct($message, $status);
    }

    /**
     * 不正リクエスト例外を取得します。
     *
     * @param string $key
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function badRequest(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_BAD_REQUEST, $key, $replace);
    }

    /**
     * 認証エラー例外を取得します。
     *
     * @param string $key エラーメッセージキー
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function unauthorized(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_UNAUTHORIZED, $key, $replace);
    }

    /**
     * アクセス拒否の例外を取得します。
     *
     * @param string $key エラーメッセージキー
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function forbidden(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_FORBIDDEN, $key, $replace);
    }

    /**
     * リソース未検出の例外を取得します。
     *
     * @param string $key エラーメッセージキー
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function notFound(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_NOT_FOUND, $key, $replace);
    }

    /**
     * リクエスト制限の例外を取得します。
     *
     * @param string $key エラーメッセージキー
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function manyRequest(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_TOO_MANY_REQUESTS, $key, $replace);
    }

    /**
     * サーバー内部エラー例外を取得します。
     *
     * @param string $key エラーメッセージキー
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function internal(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_INTERNAL_SERVER_ERROR, $key, $replace);
    }

    /**
     * サービス利用不可の例外を取得します。
     *
     * @param string $key エラーメッセージキー
     * @param array $replace メッセージ置換文字列
     * @return self 例外
     */
    public static function unavailable(string $key, array $replace = []): self
    {
        return new self(Response::HTTP_SERVICE_UNAVAILABLE, $key, $replace);
    }
}

File: backends/java-spring/src/main/java/com/authorization/support/exceptions/AppException.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.support.exceptions;

/**
 * アプリケーションExceptionクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
public class AppException extends RuntimeException {

    private final int statusCode;

    /**
     * コンストラクタ。
     *
     * @param statusCode HTTPステータスコード
     * @param message メッセージキー
     */
    public AppException(int statusCode, String message) {
        super(message);
        this.statusCode = statusCode;
    }

    public int getStatusCode() {
        return statusCode;
    }

    /**
     * 不正リクエスト例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException badRequest(String message) {
        return new AppException(400, message);
    }

    /**
     * 認証エラー例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException unauthorized(String message) {
        return new AppException(401, message);
    }

    /**
     * アクセス拒否の例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException forbidden(String message) {
        return new AppException(403, message);
    }

    /**
     * リソース未検出の例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException notFound(String message) {
        return new AppException(404, message);
    }

    /**
     * リクエスト制限の例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException manyRequest(String message) {
        return new AppException(429, message);
    }

    /**
     * サーバー内部エラー例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException internal(String message) {
        return new AppException(500, message);
    }

    /**
     * サービス利用不可の例外を取得します。
     *
     * @param message エラーメッセージキー
     * @return 例外
     */
    public static AppException unavailable(String message) {
        return new AppException(503, message);
    }
}
